File: scala/linked-list/src/main/scala/Deque.scala
import Deque.Element

class Deque[T] {

  private var first: Option[Element[T]] = None
  private var last: Option[Element[T]] = None

  def unshift(t: T): Unit = {
    val newElement = Some(Element(t, next = first))
    first foreach (_.prev = newElement)
    first = newElement
    last = last orElse first
  }

  def push(t: T): Unit = {
    val newElement = Some(Element(t, prev = last))
    last foreach (_.next = newElement)
    last = newElement
    first = first orElse last
  }

  def shift: Option[T] = {
    val result = first map (_.value)
    first = first flatMap (_.next)
    result
  }

  def pop: Option[T] = {
    val result = last map (_.value)
    last = last flatMap (_.prev)
    result
  }

  def isEmpty: Boolean = first.isEmpty
}

object Deque {
  def apply[T]() = new Deque[T]

  case class Element[T](value: T,
      var next: Option[Element[T]] = None,
      var prev: Option[Element[T]] = None)
}

File: java/linked-list/src/main/java/Deque.java
import java.util.Optional;

public class Deque<T> {

    Optional<Element<T>> first = Optional.empty();
    Optional<Element<T>> last = Optional.empty();

    public void push(T t) {
        Optional<Element<T>> newElement =
            Optional.of(new Element<T>(t, last, Optional.empty()));
        last.ifPresent(l -> l.next = newElement);
        last = newElement;
        first = first.isPresent() ? first : last;
    }

    public T pop() {
        Optional<T> result = last.map(l -> l.value);
        last = last.flatMap(l -> l.prev);
        return result.get(); // more natural would be to return the Optional itself
    }

    public void unshift(T t) {
        Optional<Element<T>> newElement =
           Optional.of(new Element<T>(t, Optional.empty(), first));
        first.ifPresent(f -> f.prev = newElement);
        first = newElement;
        last = last.isPresent() ? last : first;
    }

    public T shift() {
        Optional<T> result = first.map(f -> f.value);
        first = first.flatMap(f -> f.next);
        return result.get(); // again, more natural would be to return the Optional itself
    }

    private static class Element<T> {
        private T value;
        private Optional<Element<T>> prev;
        private Optional<Element<T>> next;

        private Element(T value, Optional<Element<T>> prev, Optional<Element<T>> next) {
            this.value = value;
            this.next = next;
            this.prev = prev;
        }
    }
}
